modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentLocalStore.java

```diff
@@ -231,23 +234,33 @@
 
     /** {@inheritDoc} */
     @Override public GridDeployment searchDeploymentCache(GridDeploymentMetadata meta) {
-        return deployment(meta.alias());
+        return deployment(meta.alias(), meta.classLoader());
     }
 
     /**
      * @param alias Class alias.
+     * @param clsLdr ClassLoader.
      * @return Deployment.
      */
-    @Nullable private GridDeployment deployment(String alias) {
+    @Nullable private GridDeployment deployment(String alias, ClassLoader clsLdr) {
         Deque<GridDeployment> deps = cache.get(alias);
 
         if (deps != null) {
-            GridDeployment dep = deps.peekFirst();
+            for (GridDeployment dep : deps) {
+                if (dep.classLoader() == clsLdr) {
+                    if (log.isDebugEnabled())
```

Comment:
If you used a trace, you should use the corresponding condition (`log.isTraceEnabled()`).